tests/solidity_test_helpers/ZeroX/ZeroXPoC.sol

```diff
@@ -0,0 +1,446 @@
+        return true;
+    }
+
+    function withdraw(ERC20 _token, uint256 _amount) public nonReentrant returns (bool) {
+        require(_token != ERC20(0));
+        uint256 _heldAmount = tokenBalances[_token][msg.sender];
+        require(_heldAmount >= _amount);
+        require(_token.transfer(msg.sender, _amount));
+        tokenBalances[_token][msg.sender] = _heldAmount.sub(_amount);
+        Withdraw(msg.sender, _token, _amount, tokenBalances[_token][msg.sender]);
+        return true;
+    }
+
+    // TODO could add something for depositing ETH too and credit CASH
+
+    /*
+    / Exchange functions
+    */
+
+    /// @dev Fills the input order.
+    /// @param orderAddresses Array of order's maker, market
+    /// @param orderValues Array of order's outcome, orderType, amount, price, expirationTimestampInSec, and salt.
+    /// @param fillAmount Desired amount to fill.
+    /// @param v ECDSA signature parameter v.
+    /// @param r ECDSA signature parameters r.
+    /// @param s ECDSA signature parameters s.
+    /// @return success.
+    function fillOrder(
+          address[2] orderAddresses,
+          uint[6] orderValues,
+          uint fillAmount,
+          uint8 v,
+          bytes32 r,
+          bytes32 s)
+          public
+          nonReentrant
+          returns (bool)
+    {
+        Order memory order = Order({
+            maker: orderAddresses[0],
+            market: IMarket(orderAddresses[1]),
+            outcome: orderValues[0],
+            orderType: orderValues[1],
+            amount: orderValues[2],
+            price: orderValues[3],
+            expirationTimestampInSec: orderValues[4],
+            orderHash: getOrderHash(orderAddresses, orderValues)
+        });
+
+        require(order.amount > 0 && fillAmount > 0);
+
+        require(isValidSignature(
+            order.maker,
+            order.orderHash,
+            v,
+            r,
+            s
+        ));
+
+        if (controller.getTimestamp() >= order.expirationTimestampInSec) {
+            Error(uint8(Errors.ORDER_EXPIRED), order.orderHash);
+            return false;
+        }
+
+        uint _remainingAmount = order.amount.sub(getUnavailableAmount(order.orderHash));
+        uint _toFillAmount = fillAmount.min(_remainingAmount);
+        if (_toFillAmount == 0) {
+            Error(uint8(Errors.ORDER_FULLY_FILLED_OR_CANCELLED), order.orderHash);
+            return false;
+        }
+
+        filled[order.orderHash] = filled[order.orderHash].add(_toFillAmount);
+
+        Fill(
+            msg.sender,
+            _toFillAmount,
+            order.orderHash
+        );
+
+        _toFillAmount = tradeMakerSharesForFillerShares(order, _toFillAmount);
+        _toFillAmount = tradeMakerSharesForFillerTokens(order, _toFillAmount);
+        _toFillAmount = tradeMakerTokensForFillerShares(order, _toFillAmount);
+        tradeMakerTokensForFillerTokens(order, _toFillAmount);
```

Comment:
I would get rid of this complexity in favor of requiring that orders either have shares escrowed or eth escrowed, and fills are either with shares or eth.

Reply:
Yea thats the tentative plan. Wanted this to have the most complexity wr might desire to be a proper poc and see if changes were needsd